includes/handlers/class-webmention-handler-mf2.php

```diff
@@ -0,0 +1,659 @@
+			return $this->to_plaintext( current( $mf['properties'][ $propname ] ) );
+		}
+		return $fallback;
+	}
+
+	/**
+	 * Returns ['html'] element of $v, or ['value'] or just $v, in order of availablility.
+	 *
+	 * @param $v Microformats Content.
+	 * @return mixed HTML Element if present.
+	 */
+	protected function to_html( $v ) {
+		if ( $this->is_embedded_html( $v ) ) {
+			return $v['html'];
+		} elseif ( $this->is_microformat( $v ) ) {
+			return webmention_sanitize_html( htmlspecialchars( $v['value'] ) );
+		}
+		return webmention_sanitize_html( htmlspecialchars( $v ) );
+	}
+
+	/**
+	 * Gets HTML of $propname or if not, $fallback.
+	 *
+	 * @param array       $mf Microformats JSON array.
+	 * @param $propname Property Name.
+	 * @param null|string $fallback Fallback if property not found.
+	 * @return mixed|null Value of proerty.
+	 */
+	protected function get_html( array $mf, $propname, $fallback = null ) {
+		if ( ! empty( $mf['properties'][ $propname ] ) && is_array( $mf['properties'][ $propname ] ) ) {
+			return $this->to_html( current( $mf['properties'][ $propname ] ) );
+		}
+		return $fallback;
+	}
+
+	/**
+	 * Gets the DateTime properties including published or updated, depending on params.
+	 *
+	 * @param $name string updated or published.
+	 * @param array                            $mf Microformats JSON array.
+	 * @param null|DateTimeImmutable           $fallback What to return if not a DateTime property.
+	 * @return mixed|null
+	 */
+	protected function get_datetime_property( $name, array $mf, $fallback = null ) {
+		if ( $this->has_property( $mf, $name ) ) {
+			$return = $this->get_plaintext( $mf, $name );
+		} else {
+			return $fallback;
+		}
+		try {
+			return new DateTimeImmutable( $return );
+		} catch ( Exception $e ) {
+			return $fallback;
+		}
+	}
+
+	/**
+	 * get all top-level items.
+	 *
+	 * @param array $mf_array the microformats array.
+	 * @param array an array of top level elements array.
+	 *
+	 * @return array Return the top level items in an array.
+	 */
+	protected function get_items( $mf_array ) {
+		if ( ! $this->is_microformat_collection( $mf_array ) ) {
+			return array();
+		}
+
+		// Get first item.
+		$first_item = current( $mf_array['items'] );
+
+		// Check if it is an h-feed.
+		if ( $this->is_type( $first_item, 'h-feed' ) && array_key_exists( 'children', $first_item ) ) {
+			$mf_array['items'] = $first_item['children'];
+		}
+
+		// Return entries.
+		return $mf_array['items'];
+	}
+
+	/**
+	 * Helper to find the correct h- node.
+	 *
+	 * @param array $mf The parsed microformats json.
+	 * @param string $url the retrieved url.
+	 *
+	 * @return array the h- node or false/
+	 */
+	protected function find_representative_item( $mf, $url ) {
+		$items = $this->get_items( $mf );
+		if ( ! is_array( $items ) || empty( $items ) ) {
+			return false;
+		}
+		// Iterate array
+		foreach ( $items as $item ) {
+			if ( $this->urls_match( $url, $this->get_plaintext( $item, 'url' ) ) ) {
```

Comment:
This can not work and is way to simple:

* If there is more than one URL it only checks with the first one.
* The URL has not to be in the URL array, one example is the reply context (`u-in-reply-to`)